Use more Express and remove body-parser dependency

StorytellerCZ opened the pull request:
While debugging some issues I have with my OAuth I made some improvements to the OAuth server part to utilize more of Express and remove the dependency on `body-parser` ~~and removed `oauthServiceName` function as we could get all of that via url params~~.

StorytellerCZ committed:
OAuth: Use more Express and remove body-parser dependency

StorytellerCZ commented:
The test failures in Travis are related, but I don't think the failures are valid and instead the tests might need updating. Will explore once I have more time. Not sure what to make of the failed CircleCI test.

denihs committed:
Merge branch 'devel' into feature/oauth-express

denihs commented:
Once you have an update about the tests we can merge it!

StorytellerCZ committed:
Replace parseurl with node:url

StorytellerCZ committed:
Merge remote-tracking branch 'origin/release-3.0.3' into feature/oauth-express

StorytellerCZ committed:
Merge remote-tracking branch 'origin/feature/oauth-express' into feature/oauth-express

StorytellerCZ committed:
Merge branch 'feature/update-webapp' into feature/oauth-express

StorytellerCZ committed:
Distinguish node official packages imports from regular packages

StorytellerCZ committed:
Merge remote-tracking branch 'origin/release-3.0.4' into feature/oauth-express

StorytellerCZ commented:
Initial look on the failing tests I still don't see why they are coming from WebApp, so I need to dive in further.

StorytellerCZ committed:
Improve URL parsing and get test without qs

StorytellerCZ committed:
For of loops and potential fix for url parsing error

StorytellerCZ committed:
Merge branch 'devel' into feature/oauth-express

StorytellerCZ committed:
Remove deprecated API usage around URL

StorytellerCZ commented:
Fixed the issue in Travis and now getting:
```
node:internal/process/promises:391
    triggerUncaughtException(err, true /* fromPromise */);
    ^
Error [ReferenceError]: __Tinytest is not defined
    at evaluate (evaluate at /home/travis/build/meteor/meteor/packages/test-in-console/puppeteerRunner.js:15:19, <anonymous>:0:19)
    at ExecutionContext._ExecutionContext_evaluate (/home/travis/build/meteor/meteor/dev_bundle/lib/node_modules/puppeteer/node_modules/puppeteer-core/lib/cjs/puppeteer/common/ExecutionContext.js:262:51)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async ExecutionContext.evaluate (/home/travis/build/meteor/meteor/dev_bundle/lib/node_modules/puppeteer/node_modules/puppeteer-core/lib/cjs/puppeteer/common/ExecutionContext.js:145:16)
    at async /home/travis/build/meteor/meteor/packages/test-in-console/puppeteerRunner.js:15:8
```

StorytellerCZ committed:
Add base to URL creation

StorytellerCZ committed:
Small cleanup

StorytellerCZ committed:
Update Send dependency

Seems like this got lost somewhere.

StorytellerCZ committed:
Remove removed Travis setting

denihs commented:
Some tests are failing; could you have a look at them? 
![image](https://github.com/user-attachments/assets/2b778311-2567-4fa8-ada4-88dd048f4bcf)

StorytellerCZ commented:
Yes, it is on my list.

StorytellerCZ committed:
Merge branch 'release-3.0.4' into feature/oauth-express

StorytellerCZ committed:
Few minor performance improvements

StorytellerCZ committed:
Merge branch 'devel' into feature/oauth-express

StorytellerCZ committed:
Merge branch 'devel' into feature/oauth-express

StorytellerCZ committed:
Update qs in WebApp

StorytellerCZ committed:
[Roles] Move _checkRoleName from client to common

StorytellerCZ committed:
Return old method in case params don't get detected by express

StorytellerCZ commented:
I think the tests are going to pass now. Will have to figure out later why the `params` were not detected in tests. Right now I have it so that if there is param with the service name it will get it first and the expensive old function to detect if from the string is there as a failback. I think the issue might be related to tests specifically, but right now I don't have the time to dig more deeply.

nachocodoner committed:
Merge branch 'devel' into feature/oauth-express

nachocodoner commented:
> While debugging some issues I have with my OAuth

@StorytellerCZ: How is this related with oauth fixes https://github.com/meteor/meteor/pull/13515#issuecomment-2552730925?

Just to clarify, does this PR include refactors, cleaning, and updates to modern tools like Express, or does it also address any fixes? Could you explain the OAuth server issue that is fixed in this PR? I want to replicate and verify it if any.

StorytellerCZ commented:
Over the time this has become more of a refactor to use Express more and update stuff. The original problem was that sometimes the callback passed through and went to the client without hitting the endpoints on the server.

StorytellerCZ committed:
Merge branch 'devel' into feature/oauth-express

StorytellerCZ committed:
Upgrade compression package

This adds support for brotli compression

StorytellerCZ committed:
Update qs, cookie-parser and add types for all the packages

qs changelog: https://github.com/ljharb/qs/blob/main/CHANGELOG.md#6140

StorytellerCZ commented:
Upgraded `compression` package which adds support for brotli compression (looking at the documentation I'm not sure if we have to do anything else to have it supported). Going to take this out for testing right now as I have some issues with OAuth to see if I can dig down to the main issue.

We should probably consider some settings to allow setting things like compression and potentially testing what is best for Meteor apps (if we can get better than the current defaults).
Docs on compression: https://github.com/expressjs/compression?tab=readme-ov-file#compression

StorytellerCZ committed:
Dedicated OAuth router

Added some imports, bit of styling.
Moved the `_oauth` routes to a dedicated Express router.

StorytellerCZ commented:
Still having issues with OAuth in Brave browser (started this primarily as I wanted to fix it). For some reason after redirect back to the server it fails to hit the server correctly:

https://github.com/user-attachments/assets/4bd3d94c-b6dd-426a-a4da-e021c3ecc7bb

When doing the test with Edge (the left window) it works fine. Not sure what exactly is the issue in Brave (the test was with Shield down), but probably some security setting. Brave has a lot more turned on, but we need to find exactly what it is as it is likely that more browsers will follow or users might turn it on. https://privacytests.org/ has the list of security settings comparison.
I think we should also look into modernizing the popup communication back to the app, but that is for another issue.
I will now look on the failing tests. Hopefully figuring out why those fail might provide a breakthrough.

StorytellerCZ committed:
Merge branch 'devel' into feature/oauth-express

Grubba27 reviewed (commented):
It seems ok – I would feel more comfortable if @nachocodoner takes a look at it too

StorytellerCZ committed:
Merge branch 'devel' into feature/oauth-express
